bin/node/runtime/src/lib.rs

```diff
@@ -744,7 +744,7 @@ impl_runtime_apis! {
 		}
 	}
 
-	impl pallet_contracts_rpc_runtime_api::ContractsApi<Block, AccountId, Balance> for Runtime {
+	impl pallet_contracts_rpc_runtime_api::ContractsApi<Block, AccountId, Balance, BlockNumber> for Runtime {
```

Comment:
But how do I access it from the declaration of `ContractsApi`. The `Block` type doesn't figure there... or does it?

Reply:
I actually had this idea already, but I dismissed it: I have **big** problems with items coming out of nowhere, i.e. when the developer didn't name them explicitly.

In my opinion the items should come either from imports, defined in the current file, or at least be explicitly named by the developer and must be easily greppable in the same file. This IMO eases reviews and also improves onboarding experience for new contributors, since this way it will be always clear where all the items are coming from.

However, I actually just took a shot to this for self-education and got biten by some nonsensical error.

<details>

```
error[E0053]: method `ContractsApi_rent_projection_runtime_api_impl` has an incompatible type for trait
   --> bin/node/runtime/src/lib.rs:806:1
    |
656 | / impl_runtime_apis! {
657 | |     impl sp_api::Core<Block> for Runtime {
658 | |         fn version() -> RuntimeVersion {
659 | |             VERSION
...   |
805 | |     }
806 | | }
    | | ^
    | | |
    | |_expected associated type, found u32
    |   in this macro invocation
    |
    = note: expected type `fn(&RuntimeApiImpl<__SR_API_BLOCK__, RuntimeApiImplCall>, &sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::BlockId<__SR_API_BLOCK__>, REWARD_CURVE::_sp_runtime::sp_application_crypto::sp_core::ExecutionContext, std::option::Option<REWARD_CURVE::_sp_runtime::AccountId32>, std::vec::Vec<u8>) -> std::result::Result<REWARD_CURVE::_sp_runtime::sp_application_crypto::sp_core::NativeOrEncoded<std::result::Result<pallet_contracts_primitives::RentProjection<<<__SR_API_BLOCK__ as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::BlockT>::Header as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::HeaderT>::Number>, pallet_contracts_primitives::ContractAccessError>>, <RuntimeApiImplCall as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::CallApiAt<__SR_API_BLOCK__>>::Error>`
               found type `fn(&RuntimeApiImpl<__SR_API_BLOCK__, RuntimeApiImplCall>, &sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::BlockId<__SR_API_BLOCK__>, REWARD_CURVE::_sp_runtime::sp_application_crypto::sp_core::ExecutionContext, std::option::Option<REWARD_CURVE::_sp_runtime::AccountId32>, std::vec::Vec<u8>) -> std::result::Result<REWARD_CURVE::_sp_runtime::sp_application_crypto::sp_core::NativeOrEncoded<std::result::Result<pallet_contracts_primitives::RentProjection<u32>, pallet_contracts_primitives::ContractAccessError>>, <RuntimeApiImplCall as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::CallApiAt<__SR_API_BLOCK__>>::Error>`
    = note: consider constraining the associated type `<<__SR_API_BLOCK__ as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::BlockT>::Header as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::HeaderT>::Number` to `u32` or calling a method that returns `<<__SR_API_BLOCK__ as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::BlockT>::Header as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::HeaderT>::Number`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html

error[E0308]: mismatched types
   --> bin/node/runtime/src/lib.rs:656:1
    |
656 | / impl_runtime_apis! {
657 | |     impl sp_api::Core<Block> for Runtime {
658 | |         fn version() -> RuntimeVersion {
659 | |             VERSION
...   |
805 | |     }
806 | | }
    | | ^
    | | |
    | | expected u32, found associated type
    | |_expected `std::result::Result<REWARD_CURVE::_sp_runtime::sp_application_crypto::sp_core::NativeOrEncoded<std::result::Result<pallet_contracts_primitives::RentProjection<u32>, pallet_contracts_primitives::ContractAccessError>>, <RuntimeApiImplCall as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::CallApiAt<__SR_API_BLOCK__>>::Error>` because of return type
    |   in this macro invocation
    |
    = note: expected type `std::result::Result<REWARD_CURVE::_sp_runtime::sp_application_crypto::sp_core::NativeOrEncoded<std::result::Result<pallet_contracts_primitives::RentProjection<u32>, _>>, _>`
               found type `std::result::Result<REWARD_CURVE::_sp_runtime::sp_application_crypto::sp_core::NativeOrEncoded<std::result::Result<pallet_contracts_primitives::RentProjection<<<__SR_API_BLOCK__ as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::BlockT>::Header as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::HeaderT>::Number>, _>>, _>`
    = note: consider constraining the associated type `<<__SR_API_BLOCK__ as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::BlockT>::Header as sp_api_hidden_includes_IMPL_RUNTIME_APIS::sp_api::HeaderT>::Number` to `u32`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
```
</details>

Reply:
Fair enough! I think this approach is a bit superfluous but indeed much less confusing.